src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,199 @@
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private int index;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T value) {
+            this.value = value;
+            this.prev = null;
+            this.next = null;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            // Insert at the beginning
+            newNode.next = head;
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            if (tail == null) {
+                tail = newNode;
+            }
+        } else if (index == size) {
+            // Insert at the end
+            newNode.prev = tail;
+            if (tail != null) {
+                tail.next = newNode;
+            }
+            tail = newNode;
+            if (head == null) {
+                head = newNode;
+            }
+        } else {
+            // Insert in the middle
+            Node<T> currentNode = head;
+            for (int i = 0; i < index - 1; i++) {
+                currentNode = currentNode.next;
+            }
+            newNode.next = currentNode.next;
+            newNode.prev = currentNode;
+            currentNode.next.prev = newNode;
+            currentNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        Node<T> currentNode = tail;
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> currentNode;
+        if (index <= size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        if (currentNode == null) {
+            throw new NullPointerException("Node at index " + index + " is null");
+        }
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size || size == 0) {
```

Comment:
```suggestion
        if (index < 0 || index >= size) {
```
and use separate method for this checking to avoid duplication in other methods